daemon/src/request/waiting_room.rs

```diff
@@ -189,30 +234,50 @@ impl<T, D> WaitingRoom<T, D> {
     fn transition<Prev, Next>(
         &mut self,
         matcher: impl FnOnce(SomeRequest<T>) -> Option<Prev>,
-        transition: impl FnOnce(Prev) -> Either<Request<TimedOut, T>, Next>,
+        transition: impl FnOnce(Prev) -> (Next, Event),
         urn: &Urn,
-    ) -> Result<(), Error>
+    ) -> Result<TransitionWithResult<T, ()>, Error>
     where
         T: Clone,
         Prev: Clone,
         Next: Into<SomeRequest<T>> + Clone,
     {
         match self.get(urn) {
             None => Err(Error::MissingUrn(urn.clone())),
-            Some(request) => match request.clone().transition(matcher, transition) {
-                Either::Right(Either::Right(next)) => {
-                    self.requests.insert(urn.id, next.into());
-                    Ok(())
-                },
-                Either::Right(Either::Left(timeout)) => {
-                    self.requests.insert(urn.id, timeout.clone().into());
-                    Err(timeout.into())
-                },
-                Either::Left(mismatch) => Err(Error::StateMismatch((&mismatch).into())),
+            Some(request) => {
+                let state_before = self.requests.clone();
+                match request.clone().transition(matcher, transition) {
+                    Either::Right((next, event)) => {
+                        let req = next.into();
+                        let t = req.timestamp().clone();
+                        self.requests.insert(urn.id, req);
+                        Ok(TransitionWithResult {
+                            state_before,
+                            state_after: self.requests.clone(),
+                            event,
+                            timestamp: t,
+                            result: (),
+                        })
+                    },
+                    Either::Left(mismatch) => Err(Error::StateMismatch((&mismatch).into())),
+                }
             },
         }
     }
 
+    /// Create a transition where the before and after state are the same
+    pub fn tick(&self, timestamp: T) -> Transition<T>
```

Comment:
Hmmmm I'm curious about this, as per one of my previous comments in `event.rs` :thinking: Why do you think it's needed?